sign/dilithium/mode3/internal/sample.go

```diff
@@ -1,8 +1,8 @@
 package internal
 
 import (
+	"github.com/cloudflare/circl/hash/f1600x4"
```

Comment:
Renaming the package to some standard name following FIPS202 (e.g. KeccakF1600) will be appreciated.

f1600x4_other.go ->  f1600x4_generic.go
The generic implementation must work too, for example using circl/internal/shake/keccakf.go. 
OTOH: (If we want people don't use the generic implementation, the entire package must be marked as only for amd64.)

Handle the case of AMD64 machines without support for AVX2. -- The `Available` flag delegates this case to the user, who might not handle this situation properly.

Some guidance must be given in the documentation that shows how to effectively load four states interleaved.

Reply:
> Renaming the package to some standard name following FIPS202 (e.g. KeccakF1600) will be appreciated.

I renamed it to `keccakf1600x4`.

> f1600x4_other.go -> f1600x4_generic.go
> The generic implementation must work too, for example using circl/internal/shake/keccakf.go.
> OTOH: (If we want people don't use the generic implementation, the entire package must be marked as only for amd64.)

Even on amd64, people should not use the package if AVX2 is not available.  I dislike a generic implementation as it's slower (due to transposing.) Also I would like to add an ARM optimized version.

> Handle the case of AMD64 machines without support for AVX2. -- The `Available` flag delegates this case to the user, who might not handle this situation properly.

Same argument as above: if AVX2 is not available a generic implementation would be prohibitively slow.

> Some guidance must be given in the documentation that shows how to effectively load four states interleaved.

I'll add an example.

Reply:
I've added the example.

Reply:
I'd agree with @armfazh here. Ideally there's a generic implementation, doesn't have to be fast, but it works. If not, guard the entire package with `amd64` build tags, so it's literally impossible to build against this package on other platforms.

Reply:
> If not, guard the entire package with `amd64` build tags, so it's literally impossible to build against this package on other platforms.

That doesn't really solve the issue as not every amd64 system has AVX2 to begin with.

Reply:
It is clear that the generic implementation will be slower, but it should have one because amd64 pre-AVX2 machines still need to be able to compile and use this external package.

It would be great to have build tags like machine+feature, so one can make the package  compiles only for amd64+avx2.

Reply:
Also for the generic implementation, you don't need to perform transpose every time provided you read data from the right place.

Reply:
> It is clear that the generic implementation will be slower, but it should have one because amd64 pre-AVX2 machines still need to be able to compile and use this external package.

They can use it!  A pre-AVX2 machine can compile this code just fine.  If the use is guarded by the `Available` flag, everything is just fine.  (Dilithium, for instance, runs just fine on my pre-AVX2 MacBook because it uses the `Available` flag to check before using f1600x4.

Reply:
> Also for the generic implementation, you don't need to perform transpose every time provided you read data from the right place.

But you can't view this function in isolation: for every application there is some kind of transposition going on.  The generic Dilithium code is faster than the code that uses f1600x4 if the latter has a generic implementation.  This is because the latter needs to interleave the data into the state whereas the former doesn't.